domains/client/domain-operator/src/tests.rs

```diff
@@ -1789,6 +1789,191 @@ async fn test_bundle_equivocation_fraud_proof() {
+    let directory = TempDir::new().expect("Must be able to create temporary directory");
+
+    let mut builder = sc_cli::LoggerBuilder::new("");
+    builder.with_colors(false);
+    let _ = builder.init();
+
+    let tokio_handle = tokio::runtime::Handle::current();
+
+    // Start Ferdie
+    let mut ferdie = MockConsensusNode::run(
+        tokio_handle.clone(),
+        Ferdie,
+        BasePath::new(directory.path().join("ferdie")),
+    );
+
+    // Run Alice (a evm domain authority node) with `skip_empty_bundle_production` set to `true`
+    let mut alice = domain_test_service::DomainNodeBuilder::new(
+        tokio_handle.clone(),
+        Alice,
+        BasePath::new(directory.path().join("alice")),
+    )
+    .build_evm_node(Role::Authority, GENESIS_DOMAIN_ID, &mut ferdie)
+    .await;
+    produce_blocks!(ferdie, alice, 1).await.unwrap();
+
+    // produce a transfer from alice to bob
+    // we send two transactions, first one must succeed, while second one will fail
+    // second extrinsic fails due to not enough balance but should still be included in the domain
+    // block body
+    let pre_alice_free_balance = alice.free_balance(Alice.to_account_id());
+    let alice_account_nonce = alice.account_nonce();
+
+    // first ext that must succeed
+    let tx = alice.construct_extrinsic(
+        alice_account_nonce,
+        pallet_balances::Call::transfer_allow_death {
+            dest: Bob.to_account_id(),
+            // we send half of alice balance to bob
+            value: pre_alice_free_balance / 2,
+        },
+    );
+    alice
+        .send_extrinsic(tx)
+        .await
+        .expect("Failed to send extrinsic");
+
+    // second ext that will fail due to balance restrictions.
+    let tx = alice.construct_extrinsic(
+        alice_account_nonce + 1,
+        pallet_balances::Call::transfer_allow_death {
+            dest: Bob.to_account_id(),
+            // try to send value that is more than what alice has
+            value: pre_alice_free_balance,
+        },
+    );
+    alice
+        .send_extrinsic(tx)
+        .await
+        .expect("Failed to send extrinsic");
+
+    // Produce a bundle and submit to the tx pool of the consensus node
+    let (_slot, bundle) = ferdie.produce_slot_and_wait_for_bundle_submission().await;
+    assert!(bundle.is_some());
+    let bundle = bundle.unwrap();
+    assert_eq!(bundle.extrinsics.len(), 2);
+
+    // produce block and import domain block
+    produce_blocks!(ferdie, alice, 1).await.unwrap();
+
+    // domain block body should have 3 extrinsics
+    // timestamp inherent, successful transfer, failed transfer
+    let best_hash = alice.client.info().best_hash;
+    let domain_block_extrinsics = alice.client.block_body(best_hash).unwrap();
+    assert_eq!(domain_block_extrinsics.unwrap().len(), 3);
+
+    // next bundle should have er which should a total of 5 trace roots
+    // timestamp + success ext + failed ext + finalize + final_state_root
```

Comment:
I see why you are confused here. It this comment should be updated to following to avoid confusion
`pre_timestamp_root + pre_success_ext_root + pre_failed_ext_root + pre_finalize_block_root + post_finalize_block_root`
Since we can already fetch `pre_initialize_root` from the parent's `post_finalize_block_root`

Reply:
I mean the comment should be changed to `initialize_block + timestamp + success ext + failed ext + finalize_block` since the definition of execution trace is the state root after `initialize_block/apply_extrinsic/finalize_block`

Reply:
No that is not what trace roots are holding. What trace roots are holding are pre roots of the actual excution phase. So going by you comments, it is even more confusing to me 
`initialize_block + timestamp + success ext + failed ext + finalize_block` since this does not give me any idea whether its holding pre or post state root

Reply:
>What trace roots are holding are pre roots of the actual excution phase.

It is the way how we collect them, but the `ER::execution_trace` is post root of `[initialize_block, apply_extrinsic.., finalize_block]` since in the invalid state transition fraud proof, we are proving the post root instead of the pre root

Reply:
I think we are simply talking about the same things 🙂